src/bridge/mod.rs

```diff
@@ -263,16 +306,23 @@ impl NeovimRuntime {
         event_loop_proxy: EventLoopProxy<UserEvent>,
         grid_size: Option<Dimensions>,
     ) -> Result<()> {
-        let handler = start_editor(event_loop_proxy);
-        let runtime = self.runtime.as_ref().unwrap();
-        let session = runtime.block_on(launch(handler, grid_size))?;
-        runtime.spawn(run(session));
+        let handler = start_editor(event_loop_proxy.clone());
+        let session = self
+            .runtime
+            .block_on(launch(handler, grid_size, &mut self.join_set))?;
+        self.join_set
+            .spawn_on(run(session, event_loop_proxy), self.runtime.handle());
         Ok(())
     }
 }
 
 impl Drop for NeovimRuntime {
     fn drop(&mut self) {
-        self.runtime.take().unwrap().shutdown_background();
+        log::info!("Starting neovim runtime shutdown");
+        let start = Instant::now();
+        shutdown_ui();
```

Comment:
Shutting down the UI explicitly from the runtime drop doesn't feel very clean. IMHO the runtime should be concerned only about itself, and shutting down other things should be done elsewhere.

As an aside, I'm not sure why the runtime is created manually instead of just using `tokio::main`. I did some refactoring earlier and eliminated the entire `NeovimRuntime` and just used `tokio::main`, and it was a lot cleaner.

Reply:
The function is probably a bit poorly named, it shuts down the UI event handlers, not the whole UI system. And we do have to wait for them here, otherwise they start to print stuff to stderr instead of the logging system, as the rest of the runtime shuts down while they continue running.